benchmarks/bench_merge.cpp

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2021 Igor Kushnir <igorkuo@gmail.com>.
+ *
+ * SPDX-License-Identifier: MIT
+ */
+#include <cstdlib>
+#include <algorithm>
+#include <chrono>
+#include <iomanip>
+#include <iostream>
+#include <string>
+#include <vector>
+#include <gfx/timsort.hpp>
+#include "benchmarker.hpp"
+
+template <typename value_t>
+struct Bench {
+    void operator()(const std::vector<value_t> &source) const {
+        const int size = static_cast<int>(source.size());
+        const std::vector<int> middle_positions = {
+            0, 1, 2, 5, 100, size/100, size/20, size/5, size/3, size/2, 3*size/4,
+            6*size/7, 24*size/25, 90*size/91, size-85, size-8, size-2, size-1, size
+        };
+
+        int result_count = 0;
+        Result result_sum = {};
+
+        std::cerr << "middle\\algorithm:\tstd::inplace_merge\ttimmerge" << std::endl;
+        constexpr int width = 10;
+        constexpr const char* padding1 = "       \t";
+        constexpr const char* padding2 = "        \t";
+
+        std::vector<value_t> a(source);
+        for (auto middle : middle_positions) {
+            if (middle < 0 || middle > size) {
+                continue;
+            }
+
+            std::copy(source.begin(), source.end(), a.begin());
+            std::sort(a.begin(), a.begin() + middle);
+            std::sort(a.begin() + middle, a.end());
+            const auto result = run(a, middle);
+
+            ++result_count;
+            result_sum.first += result.first;
+            result_sum.second += result.second;
+
+            std::cerr << std::setw(width) << middle
+                      << padding1 << std::setw(width) << result.first
+                      << padding2 << std::setw(width) << result.second
+                      << std::endl;
+        }
+
+        std::cerr << "   AVERAGE"
+                  << padding1 << std::setw(width) << result_sum.first / result_count
+                  << padding2 << std::setw(width) << result_sum.second / result_count
+                  << std::endl;
+    }
+
+private:
+    using Result = std::pair<double, double>;
+
+    Result run(const std::vector<value_t> &a, const int middle) const {
+        std::vector<value_t> b(a);
```

Comment:
You can just take `a` by value to avoid copying it here.

Reply:
No, I can't, because the loop below copies repeatedly from the original `a` to `b`:

```cpp
std::copy(a.begin(), a.end(), b.begin());
```

Reply:
Changed to `std::vector<value_t> b(a.size());` here and in one more place in *bench_merge.cpp*.